src/runtime/pyscope.cs

```diff
@@ -0,0 +1,427 @@
+﻿using System;
+using System.Dynamic;
+
+namespace Python.Runtime
+{
+    public class PyScopeException : Exception
+    {
+        public PyScopeException(string message)
+            : base(message)
+        {
+
+        }
+    }
+
+    /// <summary>
+    /// Classes implement this interface must be used with GIL obtained.
+    /// </summary>
+    public interface IPyObject : IDisposable
+    {
+    }
+
+    public class PyScope : DynamicObject, IPyObject
+    {
+        public string Name
+        {
+            get;
+            private set;
+        }
+
+        private bool isDisposed;
+
+        internal PyScope(string name)
+        {
+            this.Name = name;
+            variables = Runtime.PyDict_New();
+            if (variables == IntPtr.Zero)
+            {
+                throw new PythonException();
+            }
+            Runtime.PyDict_SetItemString(
+                variables, "__builtins__",
+                Runtime.PyEval_GetBuiltins()
+            );
+            InitVariables();
+        }
+
+        //To compitable with the python module.
+        //Enable locals() and glocals() usage in the python script.
+        private void InitVariables()
+        {
+            SetVariable("locals", (Func<PyDict>)Variables);
+            SetVariable("globals", (Func<PyDict>)Variables);
+        }
+        
+        /// <summary>
+        /// the dict for scope variables
+        /// </summary>
+        internal IntPtr variables { get; private set; }
+
+        public long Refcount
+        {
+            get
+            {
+                return Runtime.Refcount(variables);
+            }
+        }
+
+        public event Action<PyScope> OnDispose;
+
+        public PyDict Variables()
+        {
+            Runtime.XIncref(variables);
+            return new PyDict(variables);
+        }
+
+        public PyScope CreateScope()
+        {
+            var scope = new PyScope(null);
+            scope.ImportScope(this);
+            return scope;
+        }
+
+        public void ImportScope(string name)
```

Comment:
I'm not sure about this name. I think here the distinction in global/local would help clarifying the semantics: Variables of the "imported" scope should be treated as global.